gt/main.c

```diff
@@ -125,30 +140,44 @@ gt_parse_incoming_pkt(struct rte_mbuf *pkt, struct gt_packet_headers *info)
 	if (pkt->data_len < parsed_len + sizeof(struct ipv4_hdr) + 4)
 		return -1;
 
-	if (outer_ipv4_hdr != NULL)
- 		inner_ipv4_hdr = (struct ipv4_hdr *)&outer_ipv4_hdr[1];
-	else
- 		inner_ipv4_hdr = (struct ipv4_hdr *)&outer_ipv6_hdr[1];
+	if (outer_ipv4_hdr != NULL) {
+		inner_ipv4_hdr =
+			(struct ipv4_hdr *)ipv4_skip_exthdr(outer_ipv4_hdr);
+	} else {
+		inner_ipv4_hdr = (struct ipv4_hdr *)((uint8_t *)outer_ipv6_hdr
+			+ outer_ipv6_hdr_len);
+	}
 
  	inner_ip_ver = (inner_ipv4_hdr->version_ihl & 0xF0) >> 4;
 	info->inner_l3_hdr = inner_ipv4_hdr;
 
 	if (inner_ip_ver == 4) {
 		info->inner_ip_ver = ETHER_TYPE_IPv4;
 		info->l4_proto = inner_ipv4_hdr->next_proto_id;
-		info->l4_hdr = &inner_ipv4_hdr[1];
+		info->l4_hdr = (void *)ipv4_skip_exthdr(inner_ipv4_hdr);
 	} else if (likely(inner_ip_ver == 6)) {
+		uint8_t nexthdr;
+		int l4_offset;
+
 		/*
-	 	 * Make sure that the packet has space for
+		 * Make sure that the packet has space for
 		 * at least 4 bytes for the l4 header.
-	 	 */
+		 */
 		if (pkt->data_len < parsed_len + sizeof(struct ipv6_hdr) + 4)
 			return -1;
 
 		inner_ipv6_hdr = (struct ipv6_hdr *)info->inner_l3_hdr;
+		l4_offset = ipv6_skip_exthdr(inner_ipv6_hdr, pkt->data_len -
+			parsed_len, &nexthdr);
+                if (l4_offset < 0) {
+                        RTE_LOG(ERR, GATEKEEPER,
+                                "gt: failed to parse the IPv6 packet's extension headers!\n");
+			return -1;
+                }
+
 		info->inner_ip_ver = ETHER_TYPE_IPv6;
-		info->l4_proto = inner_ipv6_hdr->proto;
-		info->l4_hdr = &inner_ipv6_hdr[1];
+		info->l4_proto = nexthdr;
+		info->l4_hdr = (void *)(inner_ipv6_hdr + l4_offset);
```

Comment:
Bug: variable `inner_ipv6_hdr` if of the type `struct ipv6_hdr *`, so adding `l4_offset` is moving `l4_offset * sizeof(struct ipv6_hdr)` bytes instead of `l4_offset` bytes.